### ISSUE TITLE
AngularJS project written in JavaScript, bundled with Webpack

### ISSUE THREAD

**Issue by adrianmarinica** [monounity/karma-typescript#161]
Hello!

We've been using karma-typescript for the past couple of months with an AngularJS project written in JavaScript. Things worked great when testing individual files (controllers, services, filters, directives etc.). 

This week we migrated to Webpack and the tests are failing with an obvious message, `Uncaught ReferenceError: module is not defined`. I say obvious because of course the browser has no understanding of the `module` in `module.exports`. The less obvious part is how do we get around this issue?

I've seen issue #27 and tried to follow the example in https://github.com/monounity/karma-typescript/tree/master/examples/angularjs/app but both those pieces of information are based on an AngularJS application written in TypeScript. My problem is that I don't know how to configure karma with karma-typescript and still be able to inject / mock controllers, services etc in the specs with my JavaScript application. Is it possible?

I like karma-typescript very much because it doesn't require me to transpile the tests in JavaScript. Maybe someone can point me in the right direction? 

If you need more information from me, just let me know.

Thanks a bunch!

**Comment by erikbarke** [comment]
Hey @adrianmarinica, what does your karma config look like? Do you have a repo I could take a look at?

**Comment by adrianmarinica** [comment]
Hey @erikbarke, thanks for the response! I managed to recreate the issue in a minimalistic [repository](https://github.com/adrianmarinica/karma-typescript-angularjs-sample).

If you want to see the issue, you just have to `npm install`, `npm start` (for webpack) and `npm test`. 

Looking forward to your input. Thanks!

**Comment by erikbarke** [comment]
This is what I've got this far:

tsconfig.json:
```js
{
  "compilerOptions": {
    "allowJs": true,
    "module": "commonjs",
    "noImplicitAny": true,
    "removeComments": true,
    "preserveConstEnums": true,
    "sourceMap": true
  },
  "include": [
    "app"
  ],
  "compileOnSave": false
}
```

karma.conf.js:
```js
module.exports = function (config) {
    config.set({

        frameworks: ["jasmine", "karma-typescript"],

        files: [
            "app/**/*.+(js|ts)"
        ],

        preprocessors: {
            "app/**/*.+(js|ts)": ["karma-typescript"]
        },

        reporters: ["spec", "karma-typescript"],

        karmaTypescriptConfig: {
            tsconfig: "./tsconfig.json",
        },
    
        autoWatch: true,
        browsers: ["Chrome"]
    });
};
```

This works, kind of... Angular can't find the `weather` service, even though it's been registered in `app.js`.

However, If I add
```js
let app = angular.module("app", []);
require("../src/service")(app);
```
before the first `beforeEach` in the unit test, it works.

Mocking the module with `angular.mock.module("app")` seems to be connected to the module registered in `app.js`; if I change either one the test run breaks with "Module 'xxx' is not available", so I wonder why Angular can't inject the service.

**Comment by adrianmarinica** [comment]
Thanks a lot for your suggestions! I managed to integrate them and it's working great for services.

Controllers on the other hand, not so much. I think it's the same root cause that you described in your answer above. I get a classic AngularJS error, `Error: [ng:areq] Argument 'weatherController' is not a function, got undefined`. Any idea how I could trick it to make it work?

I've pushed the commits with the changes to the [repository](https://github.com/adrianmarinica/karma-typescript-angularjs-sample). If you have time, maybe you can take a look?

Thank you so much for your help!

**Comment by erikbarke** [comment]
It runs just fine now with these changes:

```diff
diff --git a/app/app.js b/app/app.js
index 2960ea3..6d3ee22 100644
--- a/app/app.js
+++ b/app/app.js
@@ -1,4 +1,5 @@
 ﻿var angular = require('angular');
+require("angular-route");
 
 var app = angular.module('app', ['ngRoute']);
 
diff --git a/app/test/controller.spec.ts b/app/test/controller.spec.ts
index ab92ca0..307947b 100644
--- a/app/test/controller.spec.ts
+++ b/app/test/controller.spec.ts
@@ -8,9 +8,6 @@ describe('weather controller', () => {
 
     var testDateValue = '2017/8/22';
 
-    let app = module("app", []);
-    require('../src/controller')(app);
-
     beforeEach(mock.module('app'));
 
     beforeEach(mock.inject((_$controller_: any, _$rootScope_: any) => {
diff --git a/app/test/service.spec.ts b/app/test/service.spec.ts
index 5759ba0..6cdc18d 100644
--- a/app/test/service.spec.ts
+++ b/app/test/service.spec.ts
@@ -4,10 +4,6 @@ import { IModule, mock, module } from "angular";
 describe('weather service', () => {
     var weather: any;
 
-    let app = module("app", []);
-    require('../src/weather-service')(app);
-    require('../src/temperature-service')(app);
-
     beforeEach(mock.module('app'));
 
     beforeEach(inject((_weather_: angular.IModule) => {
diff --git a/karma.conf.js b/karma.conf.js
index 057367c..7ac2703 100644
--- a/karma.conf.js
+++ b/karma.conf.js
@@ -12,7 +12,6 @@
 
         // list of files / patterns to load in the browser
         files: [
-            "app/vendor.bundle.js",
             "app/**/*.+(js|ts)"
         ],
```
I simply removed the hacky "let app = module("app", []);" fixes and now it works, no warnings or errors ☺️

**Comment by adrianmarinica** [comment]
Hey @erikbarke, I finally got around to trying out your solution, sorry for the delay. Everything works perfectly now in our production app, thanks a lot for the huge support!

If you want to, feel free to reference the repository I created for this issue as a sample for anybody that wants to try out karma-typescript with AngularJS and Webpack. I plan on adding a readme to it, so it will be easier to use.

Thanks again!
